signer/src/util.rs

```diff
@@ -0,0 +1,258 @@
+//! General utilities for the signer.
+
+use std::{
+    cmp::min,
+    future::Future,
+    ops::Deref,
+    sync::{
+        atomic::{AtomicU8, AtomicUsize, Ordering},
+        Arc,
+    },
+};
+
+use crate::error::Error;
+
+/// A fallback-wrapper that can failover to other clients if the current client fails.
+pub struct ApiFallbackClient<T> {
+    inner: Arc<InnerApiFallbackClient<T>>,
+}
+
+impl<T> Deref for ApiFallbackClient<T> {
+    type Target = InnerApiFallbackClient<T>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<T> Clone for ApiFallbackClient<T> {
+    fn clone(&self) -> Self {
+        Self { inner: Arc::clone(&self.inner) }
+    }
+}
+
+/// Inner implementation of the fallback client.
+pub struct InnerApiFallbackClient<T> {
+    inner_clients: Vec<T>,
+    last_client_index: AtomicUsize,
+    retry_count: AtomicU8,
+}
+
+impl<T> InnerApiFallbackClient<T> {
+    /// Set the number of retries to perform before giving up.
+    pub fn set_retry_count(&self, retry_count: u8) {
+        self.retry_count.store(retry_count, Ordering::Relaxed);
+    }
+
+    /// Get a reference to the current inner API client.
+    pub fn get_client(&self) -> &T {
+        &self.inner_clients[self.last_client_index.load(Ordering::Relaxed)]
+    }
+
+    /// Execute a closure on the current client, falling back to remaining clients
+    /// if the closure returns an error.
+    pub async fn exec<'a, R, E, F>(&'a self, f: impl Fn(&'a T) -> F) -> Result<R, Error>
+    where
+        E: std::error::Error + std::fmt::Debug,
+        E: Into<Error>,
+        F: Future<Output = Result<R, E>> + 'a,
+    {
+        let retry_count = self.retry_count.load(Ordering::Relaxed);
+        for i in 0..retry_count {
+            let client_index = self.last_client_index.load(Ordering::Relaxed);
+            let result = f(&self.inner_clients[client_index]).await;
+
+            if let Err(error) = result {
+                tracing::warn!(%error, retry_num=i, max_retries=retry_count, "failover client call failed");
+                self.last_client_index.store(
+                    (client_index + 1) % self.inner_clients.len(),
+                    Ordering::Relaxed,
+                );
+                continue;
+            }
+
+            return result.map_err(Into::into);
+        }
+
+        Err(Error::AllFailoverClientsFailed)
+    }
+}
+
+impl<T> ApiFallbackClient<T> {
+    /// Create a new fallback client from a list of clients.
+    pub fn new(clients: Vec<T>) -> Self {
```

Comment:
We probably want to check that the vector is non-empty too.

Reply:
Yeah I'll add that. In our case it's already guarded by the config validation, but yeah.

Reply:
Oh yeah that's right.

Reply:
Added in [7574dd2](https://github.com/stacks-network/sbtc/pull/549/commits/7574dd2a9d48a462426cf8874308485add831a29)